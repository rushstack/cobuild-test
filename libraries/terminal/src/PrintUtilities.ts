// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import * as tty from 'tty';
import wordwrap from 'wordwrap';
import { ITerminal } from '@rushstack/node-core-library';

/**
 * A sensible fallback column width for consoles.
 *
 * @public
 */
export const DEFAULT_CONSOLE_WIDTH: number = 80;

/**
 * A collection of utilities for printing messages to the console.
 *
 * @public
 */
export class PrintUtilities {
  /**
   * Returns the width of the console, measured in columns
   */
  public static getConsoleWidth(): number | undefined {
    const stdout: tty.WriteStream = process.stdout as tty.WriteStream;
    if (stdout && stdout.columns) {
      return stdout.columns;
    }
  }

  /**
   * Applies word wrapping.  If maxLineLength is unspecified, then it defaults to the
   * console width.
   */
  public static wrapWords(text: string, maxLineLength?: number, indent?: number): string {
    if (!indent) {
      indent = 0;
    }

    if (!maxLineLength) {
      maxLineLength = PrintUtilities.getConsoleWidth() || DEFAULT_CONSOLE_WIDTH;
    }

    // Apply word wrapping and the provided indent, while also respecting existing newlines
    // and prefix spaces that may exist in the text string already.
    const lines: string[] = text.split(/\r?\n/);
    const wrappedLines: string[] = lines.map((line) => {
      const startingSpace: RegExpMatchArray | null = line.match(/^ +/);
      const addlIndent: number = startingSpace?.[0]?.length || 0;

      if (addlIndent > 0) {
        line = line.replace(/^ +/, '');
      }

      return wordwrap(indent! + addlIndent, maxLineLength! - indent! - addlIndent, { mode: 'soft' })(line);
    });

    return wrappedLines.join('\n');
  }

  /**
   * Displays a message in the console wrapped in a box UI.
   *
   * @param boxWidth - The width of the box, defaults to half of the console width.
   */
  public static printMessageInBox(message: string, terminal: ITerminal, boxWidth?: number): void {
    if (!boxWidth) {
      const consoleWidth: number = PrintUtilities.getConsoleWidth() || DEFAULT_CONSOLE_WIDTH;
      boxWidth = Math.floor(consoleWidth / 2);
    }
    const maxLineLength: number = boxWidth - 10;
    const wrappedMessage: string = PrintUtilities.wrapWords(message, maxLineLength);
    const wrappedMessageLines: string[] = wrappedMessage.split('\n');

    // ╔═══════════╗
    // ║  Message  ║
    // ╚═══════════╝
    terminal.writeLine(` ╔${'═'.repeat(boxWidth - 2)}╗ `);
    for (const line of wrappedMessageLines) {
      const trimmedLine: string = line.trim();
      const padding: number = boxWidth - trimmedLine.length - 2;
      const leftPadding: number = Math.floor(padding / 2);
      const rightPadding: number = padding - leftPadding;
      terminal.writeLine(` ║${' '.repeat(leftPadding)}${trimmedLine}${' '.repeat(rightPadding)}║ `);
    }
    terminal.writeLine(` ╚${'═'.repeat(boxWidth - 2)}╝ `);
  }
}
