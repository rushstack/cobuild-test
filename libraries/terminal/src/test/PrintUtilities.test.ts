// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { StringBufferTerminalProvider, Terminal } from '@rushstack/node-core-library';

import { PrintUtilities } from '../PrintUtilities';

describe(PrintUtilities.name, () => {
  let terminalProvider: StringBufferTerminalProvider;
  let terminal: Terminal;

  beforeEach(() => {
    terminalProvider = new StringBufferTerminalProvider(false);
    terminal = new Terminal(terminalProvider);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  function validateOutput(expectedWidth: number): void {
    const outputLines: string[] = terminalProvider
      .getOutput({ normalizeSpecialCharacters: true })
      .split('[n]');
    expect(outputLines).toMatchSnapshot();

    expect(outputLines[0].trim().length).toEqual(expectedWidth);
  }

  describe(PrintUtilities.wrapWords.name, () => {
    it('respects spaces and newlines in a pre-formatted message', () => {
      const userMessage: string = [
        'An error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
        '',
        '    common/scripts/setup.sh',
        ''
      ].join('\n');

      const result: string = PrintUtilities.wrapWords(userMessage, 50, 4);
      expect(result.split(/\n/)).toMatchSnapshot();
    });

    it('applies pre-existing indents on both margins', () => {
      const message: string = [
        'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.',
        '',
        '    Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.',
        '',
        'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa.'
      ].join('\n');

      const result: string = PrintUtilities.wrapWords(message, 50);
      expect(result.split(/\n/)).toMatchSnapshot();
    });
  });

  describe(PrintUtilities.printMessageInBox.name, () => {
    const MESSAGE: string =
      'Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.';

    it('prints a long message wrapped in narrow box', () => {
      PrintUtilities.printMessageInBox(MESSAGE, terminal, 20);
      validateOutput(20);
    });

    it('prints a long message wrapped in a wide box', () => {
      PrintUtilities.printMessageInBox(MESSAGE, terminal, 300);
      validateOutput(300);
    });

    it('prints a long message wrapped in a box using the console width', () => {
      jest.spyOn(PrintUtilities, 'getConsoleWidth').mockReturnValue(65);

      PrintUtilities.printMessageInBox(MESSAGE, terminal);
      validateOutput(32);
    });

    it('respects spaces and newlines in a pre-formatted message', () => {
      const userMessage: string = [
        'An error occurred while pushing commits to git remote. Please make sure you have installed and enabled git lfs. The easiest way to do that is run the provided setup script:',
        '',
        '    common/scripts/setup.sh',
        ''
      ].join('\n');

      PrintUtilities.printMessageInBox(userMessage, terminal, 50);
      validateOutput(50);
    });
  });
});
