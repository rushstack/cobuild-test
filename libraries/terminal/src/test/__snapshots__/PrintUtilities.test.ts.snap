// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrintUtilities printMessageInBox prints a long message wrapped in a box using the console width 1`] = `
Array [
  " ╔══════════════════════════════╗ ",
  " ║    Lorem ipsum dolor sit     ║ ",
  " ║      amet, consectetuer      ║ ",
  " ║       adipiscing elit.       ║ ",
  " ║      Maecenas porttitor      ║ ",
  " ║     congue massa. Fusce      ║ ",
  " ║      posuere, magna sed      ║ ",
  " ║     pulvinar ultricies,      ║ ",
  " ║         purus lectus         ║ ",
  " ║    malesuada libero, sit     ║ ",
  " ║      amet commodo magna      ║ ",
  " ║       eros quis urna.        ║ ",
  " ╚══════════════════════════════╝ ",
  "",
]
`;

exports[`PrintUtilities printMessageInBox prints a long message wrapped in a wide box 1`] = `
Array [
  " ╔══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╗ ",
  " ║                                              Lorem ipsum dolor sit amet, consectetuer adipiscing elit. Maecenas porttitor congue massa. Fusce posuere, magna sed pulvinar ultricies, purus lectus malesuada libero, sit amet commodo magna eros quis urna.                                               ║ ",
  " ╚══════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════════╝ ",
  "",
]
`;

exports[`PrintUtilities printMessageInBox prints a long message wrapped in narrow box 1`] = `
Array [
  " ╔══════════════════╗ ",
  " ║      Lorem       ║ ",
  " ║      ipsum       ║ ",
  " ║    dolor sit     ║ ",
  " ║      amet,       ║ ",
  " ║   consectetuer   ║ ",
  " ║    adipiscing    ║ ",
  " ║      elit.       ║ ",
  " ║     Maecenas     ║ ",
  " ║    porttitor     ║ ",
  " ║      congue      ║ ",
  " ║      massa.      ║ ",
  " ║      Fusce       ║ ",
  " ║     posuere,     ║ ",
  " ║    magna sed     ║ ",
  " ║     pulvinar     ║ ",
  " ║    ultricies,    ║ ",
  " ║      purus       ║ ",
  " ║      lectus      ║ ",
  " ║    malesuada     ║ ",
  " ║     libero,      ║ ",
  " ║     sit amet     ║ ",
  " ║     commodo      ║ ",
  " ║      magna       ║ ",
  " ║    eros quis     ║ ",
  " ║      urna.       ║ ",
  " ╚══════════════════╝ ",
  "",
]
`;

exports[`PrintUtilities printMessageInBox respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  " ╔════════════════════════════════════════════════╗ ",
  " ║    An error occurred while pushing commits     ║ ",
  " ║      to git remote. Please make sure you       ║ ",
  " ║    have installed and enabled git lfs. The     ║ ",
  " ║       easiest way to do that is run the        ║ ",
  " ║             provided setup script:             ║ ",
  " ║                                                ║ ",
  " ║            common/scripts/setup.sh             ║ ",
  " ║                                                ║ ",
  " ╚════════════════════════════════════════════════╝ ",
  "",
]
`;

exports[`PrintUtilities wrapWords applies pre-existing indents on both margins 1`] = `
Array [
  "Lorem ipsum dolor sit amet, consectetuer",
  "adipiscing elit. Maecenas porttitor congue massa.",
  "",
  "    Lorem ipsum dolor sit amet, consectetuer",
  "    adipiscing elit. Maecenas porttitor",
  "    congue massa.",
  "",
  "Lorem ipsum dolor sit amet, consectetuer",
  "adipiscing elit. Maecenas porttitor congue massa.",
]
`;

exports[`PrintUtilities wrapWords respects spaces and newlines in a pre-formatted message 1`] = `
Array [
  "    An error occurred while pushing commits",
  "    to git remote. Please make sure you have",
  "    installed and enabled git lfs. The",
  "    easiest way to do that is run the",
  "    provided setup script:",
  "    ",
  "        common/scripts/setup.sh",
  "    ",
]
`;
